Update gradle version to latest

shobhitagarwal1612 opened the pull request:
Closes #2961 

<!-- 
Thank you for contributing to ODK Collect!

Before sending this PR, please read
https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md
-->

#### What has been done to verify that this works as intended?
The project builds and gradle tasks are working as usual. Also tried running the app on a device and it works as expected.

#### Why is this the best possible solution? Were any other approaches considered?
Should be updated to make use of the latest features released which improve build performance.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
Shouldn't affect users

#### Do we need any specific form for testing your changes? If so, please attach one.
Not needed

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

shobhitagarwal1612 committed:
Bump gradle to 3.3.2

shobhitagarwal1612 committed:
Configure spotbugs only for debug build

This is to prevent creation or execution of tasks for other builds to speed up build time. As recommended by gradle, either we can delay that by using Lazy task configuration. But since we only make use of debug type, so removing the config for other builds is the better option here

shobhitagarwal1612 committed:
Use fully qualified path when importing string resource

This was offsetting lint check

shobhitagarwal1612 committed:
spotbugs detections: Make inner classes static

huangyz0918 commented on config/quality.gradle:
```diff
@@ -34,42 +34,40 @@ apply plugin: 'com.github.spotbugs'
 
 spotbugs.toolVersion = '3.1.3'
 
-android.applicationVariants.all { variant ->
```
For line 37 to 72, you just improved the code format? It's good, but I think unnecessary changes may confuse reviewers, please note in the descriptions if necessary.

shobhitagarwal1612 replied:
I have written the description in the commit

huangyz0918 commented on collect_app/src/androidTest/java/org/odk/collect/android/AllWidgetsFormTest.java:
```diff
@@ -1104,7 +1104,7 @@ private String randomIntegerSeparator() {
 
     //region Custom TestRule.
-    private class FormEntryActivityTestRule extends IntentsTestRule<FormEntryActivity> {
+    private static class FormEntryActivityTestRule extends IntentsTestRule<FormEntryActivity> {
```
Is those changes in Tests related to the update of gradle version?

shobhitagarwal1612 replied:
The build was failing because of these spotbugs detections. It is possible that these are linked to the updated gradle version

lognaturel reviewed (approved):
This looks great to me with good commit messages. I'd recommend waiting until 1.21 is out, though, so it's one less change to worry about. Even if it's not user-facing, there's always some risk that things change in unpredictable ways.

lognaturel commented:
@shobhitagarwal1612 I just noticed the tests are failing. Is that related to the change here or does re-running them maybe work?

shobhitagarwal1612 commented:
I ran it locally before pushing and it was fine. Let me check what's happening on CircleCI

shobhitagarwal1612 committed:
Defer gradle tasks creation with new gradle API

https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#sec:old_vs_new_configuration_api_overview

shobhitagarwal1612 committed:
Remove redundant Xlint option while assembling any build

This resolves gradle warnings as well as speeds up build

We already have 4 static code analysis tools with proper configurations and reporting. All of the checks included in Xlint are already covered by these 4 tools. Moreover, Xlint reports in the console rather than in the xml or html format.

https://www.javaworld.com/article/2073587/javac-s--xlint-options.html

shobhitagarwal1612 committed:
Fix path of manifest file for the latest gradle API

shobhitagarwal1612 commented:
Remaining warnings are caused by the current crashlytics plugin. 
Verified by running with `-Pandroid.debug.obsoleteApi=true` flag

https://github.com/firebase/firebase-android-sdk/issues/198

grzesiek2010 commented on config/quality.gradle:
```diff
@@ -34,8 +36,7 @@ apply plugin: 'com.github.spotbugs'
 
-task("spotbugs", type: com.github.spotbugs.SpotBugsTask, dependsOn: "assembleDebug") {
-
+tasks.register("spotbugs", SpotBugsTask) {
```
In one of the previous commits, you added `dependsOn: "assembleDebug"` and here you removed it, why?

shobhitagarwal1612 replied:
It was redundant. Even after removing the `dependsOn: "assembleDebug"`, the task `assembleDebug` gets executed (probably some internal dependency) before running the `spotbugs` task. Verified by cleaning the build dir and then running the `spotbugs` task

grzesiek2010 commented:
@lognaturel any concerns or we can merge?

lognaturel commented:
No concerns now that the tests pass!
